src/librustc/mir/visit.rs

```diff
@@ -676,20 +677,17 @@ macro_rules! make_mir_visitor {
                             place: & $($mutability)? Place<'tcx>,
                             context: PlaceContext,
                             location: Location) {
-                match place {
-                    Place::Base(place_base) => {
-                        self.visit_place_base(place_base, context, location);
-                    }
-                    Place::Projection(proj) => {
-                        let context = if context.is_mutating_use() {
-                            PlaceContext::MutatingUse(MutatingUseContext::Projection)
-                        } else {
-                            PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
-                        };
-
-                        self.visit_projection(proj, context, location);
-                    }
+                if let Some(box proj) = & $($mutability)? place.projection {
+                    let context = if context.is_mutating_use() {
+                        PlaceContext::MutatingUse(MutatingUseContext::Projection)
+                    } else {
+                        PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
+                    };
+
+                    self.visit_projection(& $($mutability)? place.base, proj, context, location);
                 }
+
+                self.visit_place_base(& $($mutability)? place.base, context, location);
```

Comment:
if I'm reading this diff right, then `context` here used to be `*UseContext::Projection` if there were any projections.

Reply:
I wonder if we can get rid of `UseContext::Projection` altogether, now that we don't need recursion for `visit_place` - but I'd leave this for another PR.